core/src/main/java/org/elasticsearch/common/settings/KeyStoreWrapper.java

```diff
@@ -0,0 +1,218 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.settings;
+
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import javax.security.auth.DestroyFailedException;
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.attribute.PosixFileAttributeView;
+import java.nio.file.attribute.PosixFilePermissions;
+import java.security.GeneralSecurityException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.lucene.util.SetOnce;
+
+/**
+ * A wrapper around a Java KeyStore which provides supplements the keystore with extra metadata.
+ *
+ * Loading a keystore has 2 phases. First, call {@link #loadMetadata(Path)}. Then call
+ * {@link #loadKeystore(char[])} with the keystore password, or an empty char array if
+ * {@link #hasPassword()} is {@code false}.
+ */
+public class KeyStoreWrapper implements Closeable {
+
+    /** The version of the metadata written before the keystore data. */
+    private static final int FORMAT_VERSION = 1;
+
+    /** The keystore type for a newly created keystore. */
+    private static final String NEW_KEYSTORE_TYPE = "PKCS12";
+
+    /** A factory necessary for constructing instances of secrets in a {@link KeyStore}. */
+    private static final SecretKeyFactory passwordFactory;
+    static {
+        try {
+            passwordFactory = SecretKeyFactory.getInstance("PBE");
```

Comment:
Ok, I made this explicit, and added the algorithm name to the metadata we store before the keystore, so that we can change this in the future.